mmdet/models/anchorfree_heads/fsaf_head.py

```diff
@@ -0,0 +1,488 @@
+        bbox_targets_list = self.images_to_levels(all_bbox_targets, num_imgs, num_levels, False)
+        bbox_locs_list = self.images_to_levels(all_bbox_locs, num_imgs, num_levels, False)
+        return (labels_list, label_weights_list, bbox_targets_list,
+                bbox_locs_list, num_total_pos, num_total_neg)
+
+    def point_target_single(self,
+                            cls_score_list,
+                            bbox_pred_list,
+                            gt_bboxes,
+                            gt_bboxes_ignore,
+                            gt_labels,
+                            img_meta,
+                            cfg
+                            ):
+        num_levels = len(self.feat_strides)
+        assert len(cls_score_list) == len(bbox_pred_list) == num_levels
+        feat_lvls = self.feat_level_select(cls_score_list, bbox_pred_list, gt_bboxes, gt_labels, cfg)
+        labels = []
+        label_weights = []
+        bbox_targets = []
+        bbox_locs = []
+        device = bbox_pred_list[0].device
+        img_h, img_w, _ = img_meta['pad_shape']
+        for lvl in range(num_levels):
+            stride = self.feat_strides[lvl]
+            norm = stride * self.norm_factor
+            inds = torch.nonzero(feat_lvls == lvl).squeeze(-1)
+            h, w = cls_score_list[lvl].size()[-2:]
+            valid_h = min(int(np.ceil(img_h / stride)), h)
+            valid_w = min(int(np.ceil(img_w / stride)), w)
+
+            _labels = torch.zeros_like(cls_score_list[lvl][0], dtype=torch.long)
+            _label_weights = torch.zeros_like(cls_score_list[lvl][0], dtype=torch.float)
+            _label_weights[:valid_h, :valid_w] = 1.
+            _bbox_targets = bbox_pred_list[lvl].new_zeros((0, 4), dtype=torch.float)
+            _bbox_locs = bbox_pred_list[lvl].new_zeros((0, 3), dtype=torch.long)
+
+            if len(inds) > 0:
+                boxes = gt_bboxes[inds, :]
+                classes = gt_labels[inds]
+                proj_boxes = boxes / stride
+                ig_boxes = xyxy2xcycwh(proj_boxes)
+                ig_boxes[:, 2:] *= cfg.ignore_scale
+                ig_boxes = xcycwh2xyxy(ig_boxes)
+                pos_boxes = xyxy2xcycwh(proj_boxes)
+                pos_boxes[:, 2:] *= cfg.pos_scale
+                pos_boxes = xcycwh2xyxy(pos_boxes)
+                ig_x1 = torch.floor(ig_boxes[:, 0]).clamp(0, w - 1).int()
+                ig_y1 = torch.floor(ig_boxes[:, 1]).clamp(0, h - 1).int()
+                ig_x2 = torch.ceil(ig_boxes[:, 2]).clamp(1, w).int()
+                ig_y2 = torch.ceil(ig_boxes[:, 3]).clamp(1, h).int()
+                pos_x1 = torch.floor(pos_boxes[:, 0]).clamp(0, w - 1).int()
+                pos_y1 = torch.floor(pos_boxes[:, 1]).clamp(0, h - 1).int()
+                pos_x2 = torch.ceil(pos_boxes[:, 2]).clamp(1, w).int()
+                pos_y2 = torch.ceil(pos_boxes[:, 3]).clamp(1, h).int()
```

Comment:
There are quite a few repeated code blocks for computing a proportional region, which can be extracted to a method, similar to https://github.com/open-mmlab/mmdetection/pull/594/files#diff-6c0c212b893524716695388f57ab71d4R7.